lib/improver/utilities/compare.py

```diff
@@ -0,0 +1,291 @@
+            reporting differences
+
+    Returns:
+        None
+    """
+    if exclude_vars is None:
+        exclude_vars = []
+    actual_vars = sorted(set(actual_ds.variables.keys()) - set(exclude_vars))
+    desired_vars = sorted(set(desired_ds.variables.keys()) - set(exclude_vars))
+
+    if actual_vars != desired_vars:
+        msg = ("different variables - "
+               f"{name} {actual_vars} {desired_vars}")
+
+        reporter(msg)
+
+    coords = set()
+    for var in desired_vars:
+        coords = coords.union(desired_ds.variables[var].dimensions)
+    metadata_vars = coords.intersection(desired_vars)
+
+    for var in desired_vars:
+        var_path = f"{name}/{var}"
+        try:
+            actual_var = actual_ds.variables[var]
+            desired_var = desired_ds.variables[var]
+        except KeyError:
+            continue
+        compare_attributes(var_path, actual_var, desired_var, reporter)
+        if var in exclude_vars:
+            pass
+        elif var in metadata_vars:
+            compare_data(var_path, actual_var, desired_var,
+                         0.0, 0.0, reporter)
+        else:
+            compare_data(var_path, actual_var, desired_var,
+                         rtol, atol, reporter)
+
+
+def compare_attributes(name, actual_ds, desired_ds, reporter):
+    """
+    Compare attributes in a netCDF dataset/group.
+
+    Args:
+        name (str): group name
+        actual_ds (netCDF.Dataset): dataset produced by test run
+        desired_ds (netCDF.Dataset): dataset considered good
+        reporter (Callable[[str], None]): callback function for
+            reporting differences
+
+    Returns:
+        None
+    """
+    actual_attrs = sorted(actual_ds.ncattrs())
+    desired_attrs = sorted(desired_ds.ncattrs())
+    # ignore history attribute - this often contain datestamps and other
+    # overly specific details
+    if "history" in actual_attrs:
+        actual_attrs.remove("history")
+    if "history" in desired_attrs:
+        desired_attrs.remove("history")
+
+    if actual_attrs != desired_attrs:
+        msg = (f"different attributes of {name} -"
+               f" {actual_attrs} {desired_attrs}")
+        reporter(msg)
+
+    for key in desired_attrs:
+        try:
+            actual_attr = actual_ds.getncattr(key)
+            desired_attr = desired_ds.getncattr(key)
+            assert isinstance(desired_attr, type(actual_attr))
+            if isinstance(desired_attr, np.ndarray):
+                if not np.array_equal(actual_attr, desired_attr):
+                    msg = (f"different attribute value {name}/{key} - "
+                           f"{actual_attr} {desired_attr}")
+                    reporter(msg)
+            elif actual_attr != desired_attr:
+                msg = (f"different attribute value {name}/{key} - "
+                       f"{actual_attr} {desired_attr}")
+                reporter(msg)
+        except KeyError:
```

Comment:
Do we expect a KeyError here?

Reply:
A KeyError will occur when `actual_ds` is missing an attribute key which exists in `desired_ds` and line 240 attempts to access it. The different attribute keys will be reported earlier on line 236, but the comparison will continue in order to try and produce a more complete list of differences (assuming the reporter function doesn't raise/exit and end the comparison early)